TLSharp.Core/TelegramClient.cs

```diff
@@ -56,22 +58,32 @@ public Session Session
             if (string.IsNullOrEmpty(apiHash))
                 throw new MissingApiConfigurationException("API_HASH");
 
-            if (store == null)
-                store = new FileSessionStore();
+            this.store = store ?? new FileSessionStore();
+            this.sessionUserId = sessionUserId;
 
             this.apiHash = apiHash;
             this.apiId = apiId;
             this.handler = handler;
             this.dcIpVersion = dcIpVersion;
 
-            session = Session.TryLoadOrCreateNew(store, sessionUserId);
+            session = Session.TryLoadOrCreateNew(this.store, sessionUserId);
             transport = new TcpTransport (session.DataCenter.Address, session.DataCenter.Port, this.handler);
         }
 
         public async Task ConnectAsync(bool reconnect = false, CancellationToken token = default(CancellationToken))
         {
             token.ThrowIfCancellationRequested();
 
+            if (!transport.IsConnected)
+            {
+                // we must recreate the session because it might track dirty information 
+                // of a connection that maybe was disconnected, reusing that session will cause errors
+                session = Session.TryLoadOrCreateNew(store, sessionUserId);
+                await transport.Connect();
+            }
+            if (!transport.IsConnected)
```

Comment:
as far as I understand this cannot happen; if the call to transport.Connect above fails, then it would throw an exception from that method

Reply:
yeah should be so